multisig/src/lib.rs

```diff
@@ -64,110 +100,147 @@ impl MultiSigContract {
+        self.confirm(request_id);
+        request_id
+    }
+
     /// Remove given request and associated confirmations.
     pub fn delete_request(&mut self, request_id: RequestId) {
+        self.valid_request(request_id);
+        let request_with_signer = self.requests.get(&request_id).expect("No such request");
         assert_eq!(
-            env::current_account_id(),
-            env::predecessor_account_id(),
-            "Predecessor account must much current account"
-        );
-        assert!(
-            self.requests.get(&request_id).is_some(),
-            "No such request: either wrong number or already confirmed"
+            request_with_signer.signer_pk,
+            env::signer_account_pk(),
+            "Only creator of request can delete"
         );
-        assert!(
-            self.confirmations.get(&request_id).is_some(),
-            "Internal error: confirmations mismatch requests"
-        );
-        self.requests.remove(&request_id);
-        self.confirmations.remove(&request_id);
+        self.remove_request(request_id);
     }
 
     fn execute_request(&mut self, request: MultiSigRequest) -> PromiseOrValue<bool> {
-        match request {
-            MultiSigRequest::Transfer {
-                receiver_id,
-                amount,
-            } => Promise::new(receiver_id).transfer(amount.into()).into(),
-            MultiSigRequest::AddKey { public_key } => Promise::new(env::current_account_id())
-                .add_access_key(
-                    public_key.into(),
-                    DEFAULT_ALLOWANCE,
-                    env::current_account_id(),
-                    "add_request,delete_request,confirm"
-                        .to_string()
-                        .into_bytes(),
-                )
-                .into(),
-            MultiSigRequest::DeleteKey { public_key } => Promise::new(env::current_account_id())
-                .delete_key(public_key.into())
-                .into(),
-            MultiSigRequest::FunctionCall {
-                contract_id,
-                method_name,
-                args,
-                deposit,
-                gas,
-            } => Promise::new(contract_id)
-                .function_call(
+        let mut promise = Promise::new(request.receiver_id.clone());
+        let num_actions = request.actions.len();
+        for action in request.actions {
+            promise = match action {
+                MultiSigRequestAction::Transfer { amount } => promise.transfer(amount.into()),
+                MultiSigRequestAction::CreateAccount => promise.create_account(),
+                MultiSigRequestAction::DeployContract { code } => {
+                    promise.deploy_contract(code.into())
+                }
+                MultiSigRequestAction::AddKey {
+                    public_key,
+                    permission,
+                } if request.receiver_id == env::current_account_id() => {
```

Comment:
I think you should be able to add `FunctionCallPermission` access keys to the multi-sig. We just need to make sure that `permission.receiver_id` doesn't equal to `env::current_account_id()`. Imaging you have multi-sig account and want to login with chat app. In order to signup for this, you need to add an access key to your multi-sig account that points at `chat` contract.

Reply:
Makes sense, I will add this. We discussed in Wallet Sync from time to time, so it's a good option to have for apps.